### PR TITLE
[BugFix][Argreduce] Fix dim=0 reduction on 3D tensors

### LINKED ISSUE #899: [BUG][Argreduce] Argmax kernel fails to compile on dim=0 for 3D tensors
# [BUG][Argreduce] Argmax kernel fails to compile on dim=0 for 3D tensors

### Symptom / Motivation
Discovered during PR #894 ([Perf][Benchmark] Add multi-dim reduction benchmarks).

When benchmarking `ArgmaxFwdOp` with `dim=0` on a 3D tensor `(4, 128, 4096)`, the TVM codegen fails with:
```
Check failed: (analyzer_->CanProveEqual(abs(source->scale), 1)) is false
```

The argreduce kernel compiles and runs correctly for `dim=1` and `dim=2` on 3D tensors, but `dim=0` triggers the compilation error. The benchmark had to remove `dim=0` cases entirely to avoid silent skips.

### Root Cause Analysis
The TVM code generation for `ArgreduceKernel` does not handle the scale/stride relationship correctly when the reduction axis is the outermost (batch) dimension of a 3D tensor. This is likely a constraint in how the kernel's index mapping handles non-contiguous memory access patterns for `dim=0`.

PR #894 focused on benchmarking — fixing kernel codegen was out of scope.

### Related Files
- `tileops/ops/reduction/argmax.py` — `ArgmaxFwdOp`
- `tileops/ops/reduction/argmin.py` — `ArgminFwdOp`
- `benchmarks/ops/bench_reduce_multidim.py` — documents the limitation in section 2 comments
- `tests/ops/test_reduce_multidim.py` — multi-dim correctness tests (reference)

## Goal
Fix the argreduce kernel so `dim=0` compiles and runs correctly on 3D+ tensors, then add dim=0 benchmark cases.

## Plan

1. Investigate the TVM `CanProveEqual(abs(source->scale), 1)` assertion in the argreduce kernel codegen path for dim=0.
2. Fix the index/scale computation to handle outermost-dimension reduction on 3D tensors.
3. Add unit tests for argmax/argmin with dim=0 on 3D tensors.
4. Re-add dim=0 benchmark cases to `bench_reduce_multidim.py`.

## Constraints
- Must not regress PR #894 or PR #812
- Must not break existing argreduce dim=-1 and dim=1/2 paths

## Acceptance Criteria
- [ ] AC-1: `ArgmaxFwdOp(dtype=float16, dim=0, keepdim=False)` compiles and runs on a 3D tensor without error
- [ ] AC-2: Unit tests pass for argmax/argmin with dim=0 on 3D tensors
- [ ] AC-3: dim=0 benchmark cases added to `bench_reduce_multidim.py` and run without error

### PR DESCRIPTION
**Pull request by lcy-seso**
## Summary

Fix the argreduce kernel so `dim=0` compiles and runs correctly on 3D+ tensors. The root cause was an unconstrained `block_m` tile size that violated the TVM `CanProveEqual(abs(source->scale), 1)` assertion when reducing along the outermost dimension.

Closes #899

## Changes

- **`tileops/kernels/reduction/argreduce/fwd.py`** — Constrain `block_m` for small-N cases to fix dim=0 index/scale computation on 3D tensors.
- **`tests/ops/test_argreduce.py`** — Add unit tests for argmax/argmin with dim=0 on 3D tensors.
- **`benchmarks/ops/bench_reduce_multidim.py`** — Re-add dim=0 benchmark cases.

## Test plan

- [x] AC-1: `ArgmaxFwdOp(dtype=float16, dim=0, keepdim=False)` compiles and runs on a 3D tensor without error
  - Evidence: `pytest tests/ops/test_argreduce.py -k '3d_dim0 or spec_dim[shape2-0-False-dtype2]'` — 14 passed
- [x] AC-2: Unit tests pass for argmax/argmin with dim=0 on 3D tensors
  - Evidence: `pytest tests/ops/test_argreduce.py` — 62 passed (full suite)
- [x] AC-3: dim=0 benchmark cases added to `bench_reduce_multidim.py` and run without error
  - Evidence: `pytest benchmarks/ops/bench_reduce_multidim.py -k argreduce` — 6 passed

## Benchmark

argreduce benchmark pytest subset: 6/6 passing; dim=0 subset: 2/2 passing

### CODE REVIEW

File: tileops/kernels/reduction/argreduce/fwd.py [hunk 1]
```diff
@@ -183,23 +182,40 @@ def __init__(
...
+        reduction dimension is small (e.g. dim=0 on a 3D tensor), causing
+        ``N_padded`` to be much larger than ``N``.
+        """
         smem_per_row = self.N_padded * torch.tensor([], dtype=self.dtype).element_size()
-        max_block_m = (48 * 1024) // smem_per_row
+        max_block_m_smem = (48 * 1024) // smem_per_row
+        threads = 128
+        # TileLang layout constraint: block_m * N_padded <= 2 * threads
+        max_block_m_layout = (2 * threads) // self.N_padded
+        max_block_m = min(max_block_m_smem, max(max_block_m_layout, 1))
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py]: Addressed in f709308.

Evidence:
- The layout guard is now scoped to small-N only (`self.N < DEFAULT_ALIGNMENT`). For large N (e.g., N_padded=4096), the guard is skipped entirely and block_m is determined solely by the shared memory budget, preserving throughput.
- Reviewer-verified: dim=-1 with N_padded=4096 correctly gets block_m=4 at HEAD f709308.
- 62/62 tests pass, 6/6 benchmarks pass.

Current HEAD: f709308

File: tileops/kernels/reduction/argreduce/fwd.py [hunk 2]
```diff
@@ -183,23 +182,40 @@ def __init__(
...
-        block_ms = [bm for bm in [1, 2, 4, 8] if bm <= max_block_m]
+        max_block_m_smem = (48 * 1024) // smem_per_row
         threads_list = [128, 256]
-        configs = list(itertools.product(block_ms, threads_list))
-        return [{"block_m": bm, "threads": t} for bm, t in configs]
+        configs = []
+        for threads in threads_list:
+            # TileLang layout constraint: block_m * N_padded <= 2 * threads
+            max_block_m_layout = (2 * threads) // self.N_padded
+            max_block_m = min(max_block_m_smem, max(max_block_m_layout, 1))
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py]: Addressed in f709308.

Evidence:
- Both `default_config` and `autotune_configs` now use the same guarded pattern: the layout constraint is only applied when `self.N < DEFAULT_ALIGNMENT` (heavy padding case). The logic is consistent across both methods.
- The regression for large N_padded is fixed in both code paths.

Current HEAD: f709308

File: tileops/kernels/reduction/argreduce/fwd.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py]: Addressed in f709308.

Evidence:
- `default_config` now guards the layout constraint behind `if self.N < DEFAULT_ALIGNMENT:`, so large N_padded values (>= alignment) skip the constraint entirely. block_m selection for large-N reductions uses only the shared memory budget, as before.
- Confirmed by Reviewer: dim=-1 configs with block_m=4 compile and run at HEAD f709308.

Current HEAD: f709308

File: tileops/kernels/reduction/argreduce/fwd.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py]: Addressed in f709308.

Evidence:
- `autotune_configs` now guards the layout constraint behind `if self.N < DEFAULT_ALIGNMENT:`. For large N_padded, the autotuner explores block_m values up to the shared memory limit (block_m=4, 8), preserving efficient configurations.
- 62/62 tests pass, 6/6 benchmarks pass at HEAD f709308.

Current HEAD: f709308

File: tileops/kernels/reduction/argreduce/fwd.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py]: Addressed in f709308.

Evidence:
- The constraint is now gated on `self.N < DEFAULT_ALIGNMENT`, exactly as suggested (gating for the small-N padding case only). For large N where N_padded == N, block_m is determined solely by the shared memory budget.
- Reviewer-verified: kernel config probe confirmed the layout guard is limited to N < 256 while large-N defaults remain unchanged.

Current HEAD: f709308

File: tileops/kernels/reduction/argreduce/fwd.py [hunk 3]
```diff
@@ -183,23 +182,40 @@ def __init__(
...
+
+        Also enforces a TileLang layout-inference constraint:
+        ``block_m * N_padded <= 2 * threads`` (each thread handles at most 2
+        elements during the shared-memory copy).  This matters when the
+        reduction dimension is small (e.g. dim=0 on a 3D tensor), causing
+        ``N_padded`` to be much larger than ``N``.
+        """
         smem_per_row = self.N_padded * torch.tensor([], dtype=self.dtype).element_size()
-        max_block_m = (48 * 1024) // smem_per_row
+        max_block_m_smem = (48 * 1024) // smem_per_row
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py]: Addressed in c9d9352.

Evidence:
- Replaced hardcoded `48 * 1024` with `SHARED_MEMORY_BUDGET_BYTES` imported from `tileops.kernels.reduction._primitives` in both `_block_config` and `autotune_configs` properties.
- Import added: `from tileops.kernels.reduction._primitives import DEFAULT_ALIGNMENT, SHARED_MEMORY_BUDGET_BYTES, align_up`

Current HEAD: c9d9352

File: tileops/kernels/reduction/argreduce/fwd.py [hunk 4]
```diff
@@ -183,23 +182,40 @@ def __init__(
...
-        return [{"block_m": bm, "threads": t} for bm, t in configs]
+        configs = []
+        for threads in threads_list:
+            # TileLang layout constraint: block_m * N_padded <= 2 * threads
+            max_block_m_layout = (2 * threads) // self.N_padded
+            max_block_m = min(max_block_m_smem, max(max_block_m_layout, 1))
+            for bm in [1, 2, 4, 8]:
+                if bm <= max_block_m:
+                    configs.append({"block_m": bm, "threads": threads})
+        return configs
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py:241]: Addressed in c9d9352.

Evidence:
- Added explicit guard: `if max_block_m_smem == 0: raise ValueError(...)` in both `_block_config` and `autotune_configs` properties.
- The ValueError provides a clear diagnostic message including `smem_per_row`, `N_padded`, `dtype`, and the budget limit.
- This ensures a deterministic, actionable error when a single row exceeds the shared memory budget, rather than silently producing an invalid config.

Current HEAD: c9d9352

File: tileops/kernels/reduction/argreduce/fwd.py [hunk 5]
```diff
@@ -183,23 +186,59 @@ def __init__(
...
         smem_per_row = self.N_padded * torch.tensor([], dtype=self.dtype).element_size()
-        max_block_m = (48 * 1024) // smem_per_row
+        max_block_m_smem = SHARED_MEMORY_BUDGET_BYTES // smem_per_row
+        if max_block_m_smem == 0:
+            raise ValueError(
+                f"A single row requires {smem_per_row} bytes of shared memory, "
+                f"which exceeds the {SHARED_MEMORY_BUDGET_BYTES}-byte budget "
+                f"(N_padded={self.N_padded}, dtype={self.dtype}). "
+                f"Reduce the reduction dimension or use a dtype with smaller element size."
+            )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py:206]: Reviewer determined this is a false positive:

this is optional DRY refactoring, not a correctness/style gate. The duplicated guard in tileops/kernels/reduction/argreduce/fwd.py:198-240 is behaviorally consistent in both paths and does not expose a user-visible defect.

Current HEAD: a0ad321

File: tests/ops/test_argreduce.py [hunk 6]
```diff
@@ -60,6 +60,22 @@ class Argreduce3DFixture(FixtureBase):
...
+            "batch, seq, hidden, dtype",
+            [
+                pytest.param(4, 8, 256, torch.float16, marks=pytest.mark.smoke),
+                pytest.param(4, 8, 256, torch.bfloat16, marks=pytest.mark.full),
+                pytest.param(4, 8, 256, torch.float32, marks=pytest.mark.full),
+            ],
+        ),
+    ]
+
+
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_argreduce.py:78]: Addressed in a0ad321.

Evidence:
- Commit a0ad321 adds `Argreduce4DDim0Fixture` with 4D shape (2,4,8,256) and 6 new test functions: `test_argmax_4d_dim0`, `test_argmax_4d_dim0_keepdim`, `test_argmin_4d_dim0`, `test_argmin_4d_dim0_keepdim`, plus 4D dim=0 params in `SpecArgreduceFixture`.
- All 74 tests pass at HEAD a0ad321.

Current HEAD: a0ad321

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
This PR crosses trust-boundary stages and should be split before merge. It combines an implementation change in `tileops/kernels/reduction/argreduce/fwd.py` with new correctness tests in `tests/ops/test_argreduce.py` and benchmark updates in `benchmarks/ops/bench_reduce_multidim.py`. Under the project rules, implementation PRs must not modify tests or benchmarks. Land the kernel bugfix first, then follow with a test/benchmark PR that validates and profiles the new dim=0 support.

**Comment by lcy-seso** [comment]
Splitting per @Ibuki-wind review: trust boundary requires separating kernel fix from test/benchmark changes.

Replacement PRs:
- Kernel fix: incoming (fix/argreduce/issue-899-kernel)
- Test + benchmark coverage: incoming (test/argreduce/issue-899-coverage, depends on kernel PR)

**Comment by lcy-seso** [comment]
Adopted. Split into two PRs per trust boundary rules:

- **#911** — Kernel fix only (`tileops/kernels/reduction/argreduce/fwd.py`)
- **#912** — Test + benchmark coverage (`tests/ops/test_argreduce.py`, `benchmarks/ops/bench_reduce_multidim.py`)

#912 depends on #911. Merge order: #911 first.
